onnxruntime/core/framework/cloud_invoker.cc

```diff
@@ -33,6 +35,143 @@ CloudEndPointInvoker::CloudEndPointInvoker(const CloudEndPointConfig& config,
+}
+
+struct ResponseBuffer {
+  ResponseBuffer() = default;
+  ~ResponseBuffer() = default;
+  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(ResponseBuffer);
+
+  using Chunk = std::unique_ptr<char[]>;
+  std::list<Chunk> chunks_;
+
+  void Fill(std::string& s) const {
+    std::for_each(chunks_.begin(), chunks_.end(), [&](const Chunk& chunk) { s.append(chunk.get()); });
+  }
+};
+
+static size_t WriteResponseCallback(void* contents, size_t size, size_t nmemb, void* userp) {
+  size_t realsize = size * nmemb;
+  auto response = reinterpret_cast<struct ResponseBuffer*>(userp);
+  response->chunks_.push_back(std::make_unique<char[]>(realsize + 1));
+  memcpy(response->chunks_.back().get(), contents, realsize);
+  response->chunks_.back()[realsize] = '\0';
+  return realsize;
+}
+
+struct CurlHandler {
+  CurlHandler() {
+    curl_ = curl_easy_init();
+    curl_easy_setopt(curl_, CURLOPT_BUFFERSIZE, 102400L);
+    curl_easy_setopt(curl_, CURLOPT_NOPROGRESS, 1L);
+    curl_easy_setopt(curl_, CURLOPT_USERAGENT, "curl/7.83.1");
+    curl_easy_setopt(curl_, CURLOPT_MAXREDIRS, 50L);
+    curl_easy_setopt(curl_, CURLOPT_FTP_SKIP_PASV_IP, 1L);
+    curl_easy_setopt(curl_, CURLOPT_TCP_KEEPALIVE, 1L);
+    curl_easy_setopt(curl_, CURLOPT_WRITEFUNCTION, WriteResponseCallback);
+  }
+  ~CurlHandler() {
+    if (curl_) {
+      curl_easy_cleanup(curl_);
+      curl_ = {};
+    }
+    if (mime1_) {
+      curl_mime_free(mime1_);
+      mime1_ = {};
+    }
+    if (headers_) {
+      curl_slist_free_all(headers_);
+      headers_ = {};
+    }
+    if (from_) {
+      curl_formfree(from_);
+      from_ = {};
+    }
+  }
+  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(CurlHandler);
+  CURL* curl_{};
+  curl_mime* mime1_{};
+  struct curl_slist* headers_{};
+  struct curl_httppost* from_{};
+  struct curl_httppost* last_{};
+};
+
+onnxruntime::Status OpenAIInvoker::Send(const CloudEndPointConfig& run_options,
+                                        const InlinedVector<std::string>& /*input_names*/,
+                                        gsl::span<const OrtValue> ort_inputs,
+                                        const InlinedVector<std::string>& /*output_names*/,
+                                        std::vector<OrtValue>& ort_outputs) const {
+  const auto auth_key_iter = run_options.find(kAzureAuthKey);
+  if (run_options.end() == auth_key_iter || auth_key_iter->second.empty()) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "auth key must be specified for openai client");
+  }
+  long verbose = 0;
+  const auto verbose_iter = run_options.find(kAzureVerbose);
+  if (run_options.end() != verbose_iter) {
+    verbose = verbose_iter->second != "0" ? 1L : 0L;
+  }
+
+  CURLcode ret{};
+  CurlHandler curl_handler;
+  ResponseBuffer response_buffer;
+
+  std::string full_auth = std::string{"Authorization: Bearer "} + auth_key_iter->second;
+  curl_handler.headers_ = curl_slist_append(curl_handler.headers_, full_auth.c_str());
```

Comment:
I would've expected CurlHandler to hide the usage of specific curl functions. Now curl related functionality is leaked outside CurlHandler. Later we should either fix this or move to libcurl++ so that other invokers can utilize the full functionality of curl in the C++ way.